Cabal/Distribution/Simple/Build.hs

```diff
@@ -206,9 +206,10 @@ buildComponent verbosity numJobs pkg_descr lbi suffixes
         -- Register the library in-place, so exes can depend
         -- on internally defined libraries.
         pwd <- getCurrentDirectory
+        abi_hash <- abiHash verbosity pkg_descr distPref lbi lib clbi
         let -- The in place registration uses the "-inplace" suffix, not an ABI hash
             installedPkgInfo = inplaceInstalledPackageInfo pwd distPref pkg_descr
-                                                           (mkAbiHash "") lib' lbi clbi
+                                                           abi_hash lib' lbi clbi
```

Comment:
Hmm, actually, then does new-build do the right thing when registering packages? Looking at the logic in ProjectBuilding, it seems that we only reregister when the monitor is invalidated; but as it tracks no files, it is only ever invalidated if we reconfigure. This means that once we register once, as we rebuild, we never reregister. Which means that the 'abi' we stored in the package description is stale. So if we wanted to be "correct", we need to nub out the abi so that the registration is stable across rebuilds.

I am not entirely convinced we shouldn't recompute the ABI every time we build. In particular, it's an orthogonal method to determining what should be in a GHC environment: when we rebuild a library and update the ABI in its registration that will automatically invalidate any packages which were built against the old ABI. That seems very useful, and a bit more robust then trying to expand/shrink the environment every time someone calls new-build. I mean, maybe there is a lot of performance to be had by dropping the ABI in this case, but it's a tradeoff! Correctness versus performance.